lib/services/role.go

```diff
@@ -778,9 +778,9 @@ func ExtractFromCertificate(cert *ssh.Certificate) ([]string, wrappers.Traits, e
 // which Teleport passes along as a *tlsca.Identity. If roles and traits do not
 // exist in the certificates, they are extracted from the backend.
 func ExtractFromIdentity(access UserGetter, identity tlsca.Identity) ([]string, wrappers.Traits, error) {
-	// For legacy certificates, fetch roles and traits from the services.User
-	// object in the backend.
-	if missingIdentity(identity) {
+	// Legacy certs are not encoded with roles or traits,
+	// so we fallback to the traits and roles in the backend.
+	if identity.Groups == nil {
```

Comment:
Should this check for length == 0 like in `missingIdentity`?
```suggestion
	if len(identity.Groups) == 0 {
```

Reply:
I intentionally changed it to nil since what we want to check is whether the certificate was encoded with roles, not whether the user has any roles. If a user somehow had empty roles but `{}` was encoded, then we don't need to fallback.

In practice a user can't have empty roles so this is functionally the same, but more expressive of the intent IMO.

Reply:
Your reasoning for changing it to nil makes sense. I think the problem is that it relies on all upstream callers initializing the slice correctly. I try to avoid making a distinction between a nil slice and a zero-length slice for this reason. E.g. some decoders will initialize slices to nil or zero length indiscriminately. Since our contract for roles/Groups properly says there is no difference between nil and zero-length (user has roles or is not initialized), it may be better if this checks len/zero so we're not relying on callers to make the distinction.